File: Swift/AdhanTests/AstronomicalTests.swift
//
//  AstronomicalTests.swift
//  Adhan
//
//  Created by Ameir Al-Zoubi on 2/21/16.
//  Copyright © 2016 Batoul Apps. All rights reserved.
//

import XCTest
@testable import Adhan

class AstronomicalTests: XCTestCase {
    
    func testSolarCoordinates() {
        
        // values from Astronomical Algorithms page 165
        
        var jd = Astronomical.julianDay(year: 1992, month: 10, day: 13)
        var solar = SolarCoordinates(julianDay: jd)
        
        var T = Astronomical.julianCentury(julianDay: jd)
        var L0 = Astronomical.meanSolarLongitude(julianCentury: T)
        var ε0 = Astronomical.meanObliquityOfTheEcliptic(julianCentury: T)
        let εapp = Astronomical.apparentObliquityOfTheEcliptic(julianCentury: T, meanObliquityOfTheEcliptic: ε0)
        let M = Astronomical.meanSolarAnomaly(julianCentury: T)
        let C = Astronomical.solarEquationOfTheCenter(julianCentury: T, meanAnomaly: M)
        let λ = Astronomical.apparentSolarLongitude(julianCentury: T, meanLongitude: L0)
        let δ = solar.declination
        let α = solar.rightAscension.unwindAngle()
        
        XCTAssertEqual(T, -0.072183436,
            accuracy: 0.00000000001)
        
        XCTAssertEqual(L0, 201.80720,
            accuracy: 0.00001)
        
        XCTAssertEqual(ε0, 23.44023,
            accuracy: 0.00001)
        
        XCTAssertEqual(εapp, 23.43999,
            accuracy: 0.00001)
        
        XCTAssertEqual(M, 278.99397,
            accuracy: 0.00001)
        
        XCTAssertEqual(C, -1.89732,
            accuracy: 0.00001)
        
        // lower accuracy than desired
        XCTAssertEqual(λ, 199.90895,
            accuracy: 0.00002)
        
        XCTAssertEqual(δ, -7.78507,
            accuracy: 0.00001)
        
        XCTAssertEqual(α, 198.38083,
            accuracy: 0.00001)
        
        // values from Astronomical Algorithms page 88
        
        jd = Astronomical.julianDay(year: 1987, month: 4, day: 10)
        solar = SolarCoordinates(julianDay: jd)
        T = Astronomical.julianCentury(julianDay: jd)
        
        let θ0 = Astronomical.meanSiderealTime(julianCentury: T)
        let θapp = solar.apparentSiderealTime
        let Ω = Astronomical.ascendingLunarNodeLongitude(julianCentury: T)
        ε0 = Astronomical.meanObliquityOfTheEcliptic(julianCentury: T)
        L0 = Astronomical.meanSolarLongitude(julianCentury: T)
        let Lp = Astronomical.meanLunarLongitude(julianCentury: T)
        let ΔΨ = Astronomical.nutationInLongitude(solarLongitude: L0, lunarLongitude: Lp, ascendingNode: Ω)
        let Δε = Astronomical.nutationInObliquity(solarLongitude: L0, lunarLongitude: Lp, ascendingNode: Ω)
        let ε = ε0 + Δε
        
        XCTAssertEqual(θ0, 197.693195,
            accuracy: 0.000001)
        
        XCTAssertEqual(θapp, 197.6922295833,
            accuracy: 0.0001)
        
        // values from Astronomical Algorithms page 148
        
        XCTAssertEqual(Ω, 11.2531,
            accuracy: 0.0001)
        
        XCTAssertEqual(ΔΨ, -0.0010522,
            accuracy:  0.0001)
        
        XCTAssertEqual(Δε, 0.0026230556,
            accuracy: 0.00001)
        
        XCTAssertEqual(ε0, 23.4409463889,
            accuracy: 0.000001)
        
        XCTAssertEqual(ε, 23.4435694444,
            accuracy: 0.00001)
    }
    
    func testAltitudeOfCelestialBody() {
        let φ = 38 + (55 / 60) + (17.0 / 3600)
        let δ = -6 - (43 / 60) - (11.61 / 3600)
        let H = 64.352133
        let h = Astronomical.altitudeOfCelestialBody(observerLatitude: φ, declination: δ, localHourAngle: H)
        XCTAssertEqual(h, 15.1249,
            accuracy: 0.0001)
    }
    
    func testTransitAndHourAngle() {
        // values from Astronomical Algorithms page 103
        let longitude = -71.0833
        let Θ = 177.74208
        let α1 = 40.68021
        let α2 = 41.73129
        let α3 = 42.78204
        let m0 = Astronomical.approximateTransit(longitude: longitude, siderealTime: Θ, rightAscension: α2)
        
        XCTAssertEqual(m0, 0.81965,
            accuracy: 0.00001)
        
        let transit = Astronomical.correctedTransit(approximateTransit: m0, longitude: longitude, siderealTime: Θ, rightAscension: α2, previousRightAscension: α1, nextRightAscension: α3) / 24
        
        XCTAssertEqual(transit, 0.81980,
            accuracy: 0.00001)
        
        let δ1 = 18.04761
        let δ2 = 18.44092
        let δ3 = 18.82742
        
        let rise = Astronomical.correctedHourAngle(approximateTransit: m0, angle: -0.5667, coordinates: Coordinates(latitude: 42.3333, longitude: longitude), afterTransit: false, siderealTime: Θ,
            rightAscension: α2, previousRightAscension: α1, nextRightAscension: α3, declination: δ2, previousDeclination: δ1, nextDeclination: δ3) / 24
        XCTAssertEqual(rise, 0.51766,
            accuracy: 0.00001)
    }
    
    func testSolarTime() {
        /*
        Comparison values generated from http://aa.usno.navy.mil/rstt/onedaytable?form=1&ID=AA&year=2015&month=7&day=12&state=NC&place=raleigh
        */
        
        let coordinates = Coordinates(latitude: 35 + 47/60, longitude: -78 - 39/60)
        let solar = SolarTime(date: date(year: 2015, month: 7, day: 12, hours: 0), coordinates: coordinates)
        
        let transit = solar.transit
        let sunrise = solar.sunrise
        let sunset = solar.sunset
        let twilightStart = solar.hourAngle(angle: -6, afterTransit: false)
        let twilightEnd = solar.hourAngle(angle: -6, afterTransit: true)
        let invalid = solar.hourAngle(angle: -36, afterTransit: true)
        XCTAssertEqual(twilightStart.timeString(), "9:38")
        XCTAssertEqual(sunrise.timeString(), "10:08")
        XCTAssertEqual(transit.timeString(), "17:20")
        XCTAssertEqual(sunset.timeString(), "24:32")
        XCTAssertEqual(twilightEnd.timeString(), "25:02")
        XCTAssertEqual(invalid.timeString(), "")
    }
    
    func testRightAscensionEdgeCase() {
        let coordinates = Coordinates(latitude: 35 + 47/60, longitude: -78 - 39/60)
        let solar = [Int](0...365).map { SolarTime(date: date(year: 2016, month: 0, day: $0, hours: 0), coordinates: coordinates) }
        for (index, time) in solar.enumerated() {
            if index > 0 {
                let previousTime = solar[index-1]
                // transit from one day to another should not differ more than one minute
                XCTAssertLessThan(fabs(time.transit - previousTime.transit), 1/60)
                
                // sunrise and sunset from one day to another should not differ more than two minutes
                XCTAssertLessThan(fabs(time.sunrise - previousTime.sunrise), 2/60)
                XCTAssertLessThan(fabs(time.sunset - previousTime.sunset), 2/60)
            }
        }
    }
    
    func testCalendricalDate() {
        // generated from http://aa.usno.navy.mil/data/docs/RS_OneYear.php for KUKUIHAELE, HAWAII
        let coordinates = Coordinates(latitude: 20 + 7/60, longitude: -155 - 34/60)
        let day1solar = SolarTime(date: date(year: 2015, month: 4, day: 2, hours: 0), coordinates: coordinates)
        let day2solar = SolarTime(date: date(year: 2015, month: 4, day: 3, hours: 0), coordinates: coordinates)
        
        let day1 = day1solar.sunrise
        let day2 = day2solar.sunrise
        
        XCTAssertEqual(day1.timeString(), "16:15")
        XCTAssertEqual(day2.timeString(), "16:14")
    }
    
    func testInterpolation() {
        
        // values from Astronomical Algorithms page 25
        
        let interpolatedValue = Astronomical.interpolate(value: 0.877366, previousValue: 0.884226, nextValue: 0.870531, factor: 4.35/24)
        XCTAssertEqual(interpolatedValue, 0.876125,
                                            accuracy: 0.000001)
        
        let i1 = Astronomical.interpolate(value: 1, previousValue: -1, nextValue: 3, factor: 0.6)
        XCTAssertEqual(i1, 2.2, accuracy: 0.000001)
    }
    
    func testAngleInterpolation() {
        let i1 = Astronomical.interpolateAngles(value: 1, previousValue: -1, nextValue: 3, factor: 0.6)
        XCTAssertEqual(i1, 2.2, accuracy: 0.000001)
        
        let i2 = Astronomical.interpolateAngles(value: 1, previousValue: 359, nextValue: 3, factor: 0.6)
        XCTAssertEqual(i2, 2.2, accuracy: 0.000001)
    }
    
    func testJulianDay() {
        /*
        Comparison values generated from http://aa.usno.navy.mil/data/docs/JulianDate.php
        */
        
        XCTAssertEqual(Astronomical.julianDay(year: 2010, month: 1, day: 2), 2455198.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2011, month: 2, day: 4), 2455596.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2012, month: 3, day: 6), 2455992.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2013, month: 4, day: 8), 2456390.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2014, month: 5, day: 10), 2456787.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2015, month: 6, day: 12), 2457185.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2016, month: 7, day: 14), 2457583.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2017, month: 8, day: 16), 2457981.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2018, month: 9, day: 18), 2458379.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2019, month: 10, day: 20), 2458776.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2020, month: 11, day: 22), 2459175.500000)
        XCTAssertEqual(Astronomical.julianDay(year: 2021, month: 12, day: 24), 2459572.500000)
        
        let jdVal = 2457215.67708333
        XCTAssertEqual(Astronomical.julianDay(year: 2015, month: 7, day: 12, hours: 4.25), jdVal, accuracy: 0.000001)
        
        var components = DateComponents()
        components.year = 2015
        components.month = 7
        components.day = 12
        components.hour = 4
        components.minute = 15
        XCTAssertEqual(components.julianDate(), jdVal, accuracy: 0.000001)
        
        XCTAssertEqual(Astronomical.julianDay(year: 2015, month: 7, day: 12, hours: 8.0), 2457215.833333, accuracy: 0.000001)
        XCTAssertEqual(Astronomical.julianDay(year: 1992, month: 10, day: 13, hours: 0.0), 2448908.5, accuracy: 0.000001)
    }
    
    func testJulianHours() {
        let j1 = Astronomical.julianDay(year: 2010, month: 1, day: 3)
        let j2 = Astronomical.julianDay(year: 2010, month: 1, day: 1, hours: 48)
        XCTAssertEqual(j1, j2)
    }
    
    func testLeapYear() {
        XCTAssertFalse(Astronomical.isLeap(year: 2015))
        XCTAssertTrue(Astronomical.isLeap(year: 2016))
        XCTAssertTrue(Astronomical.isLeap(year: 1600))
        XCTAssertTrue(Astronomical.isLeap(year: 2000))
        XCTAssertTrue(Astronomical.isLeap(year: 2400))
        XCTAssertFalse(Astronomical.isLeap(year: 1700))
        XCTAssertFalse(Astronomical.isLeap(year: 1800))
        XCTAssertFalse(Astronomical.isLeap(year: 1900))
        XCTAssertFalse(Astronomical.isLeap(year: 2100))
        XCTAssertFalse(Astronomical.isLeap(year: 2200))
        XCTAssertFalse(Astronomical.isLeap(year: 2300))
        XCTAssertFalse(Astronomical.isLeap(year: 2500))
        XCTAssertFalse(Astronomical.isLeap(year: 2600))
    }
    
    func daysSinceSolsticeTest(_ value: Int, year: Int, month: Int, day: Int, latitude: Double) {
        // For Northern Hemisphere start from December 21
        // (DYY=0 for December 21, and counting forward, DYY=11 for January 1 and so on).
        // For Southern Hemisphere start from June 21
        // (DYY=0 for June 21, and counting forward)
        
        let cal = Calendar(identifier: Calendar.Identifier.gregorian)
        let dateComponents = date(year: year, month: month, day: day)
        let dayOfYear = (cal as NSCalendar).ordinality(of: .day, in: .year, for: cal.date(from: dateComponents as DateComponents)!)
        XCTAssertEqual(Astronomical.daysSinceSolstice(dayOfYear: dayOfYear, year: dateComponents.year!, latitude: latitude), value)
    }
    
    func testDaysSinceSolstice() {
        daysSinceSolsticeTest(11, year: 2016, month: 1, day: 1, latitude: 1)
        daysSinceSolsticeTest(10, year: 2015, month: 12, day: 31, latitude: 1)
        daysSinceSolsticeTest(10, year: 2016, month: 12, day: 31, latitude: 1)
        daysSinceSolsticeTest(0, year: 2016, month: 12, day: 21, latitude: 1)
        daysSinceSolsticeTest(1, year: 2016, month: 12, day: 22, latitude: 1)
        daysSinceSolsticeTest(71, year: 2016, month: 3, day: 1, latitude: 1)
        daysSinceSolsticeTest(70, year: 2015, month: 3, day: 1, latitude: 1)
        daysSinceSolsticeTest(365, year: 2016, month: 12, day: 20, latitude: 1)
        daysSinceSolsticeTest(364, year: 2015, month: 12, day: 20, latitude: 1)
        
        daysSinceSolsticeTest(0, year: 2015, month: 6, day: 21, latitude: -1)
        daysSinceSolsticeTest(0, year: 2016, month: 6, day: 21, latitude: -1)
        daysSinceSolsticeTest(364, year: 2015, month: 6, day: 20, latitude: -1)
        daysSinceSolsticeTest(365, year: 2016, month: 6, day: 20, latitude: -1)
    }
}

File: Tests/astronomical-tests.js
var unwindAngle = adhan.Math.unwindAngle;
var timeComponents = adhan.Math.timeComponents;
var dayOfYear = adhan.Date.dayOfYear;
var julianDate = adhan.Date.julianDate;
//
// Astronomical Tests
//

function timeString(hours) {
	var comps = timeComponents(hours);
    if (isNaN(comps.hours) || isNaN(comps.minutes) || isNaN(comps.seconds)) {
        return "";
    }

    // round to the nearest minute
    var minutes = (comps.minutes + Math.round((comps.seconds)/60)).toString();
    if (minutes.length == 1) {
    	minutes = "0" + minutes;
    }
    return comps.hours + ":" + minutes;
}

QUnit.test("Solar Coordinates", function(assert) {
	// values from Astronomical Algorithms page 165

	var jd = adhan.Astronomical.julianDay(1992, 10, 13);
	var solar = new adhan.SolarCoordinates(jd);

	var T = adhan.Astronomical.julianCentury(jd);
	var L0 = adhan.Astronomical.meanSolarLongitude(T);
	var E0 = adhan.Astronomical.meanObliquityOfTheEcliptic(T);
	var Eapp = adhan.Astronomical.apparentObliquityOfTheEcliptic(T, E0);
	var M = adhan.Astronomical.meanSolarAnomaly(T);
    var C = adhan.Astronomical.solarEquationOfTheCenter(T, M);
    var Lambda = adhan.Astronomical.apparentSolarLongitude(T, L0);
    var Delta = solar.declination;
    var Alpha = unwindAngle(solar.rightAscension);

	QUnit.close(T, -0.072183436, 0.00000000001);
	QUnit.close(L0, 201.80720, 0.00001);
	QUnit.close(E0, 23.44023, 0.00001);
	QUnit.close(Eapp, 23.43999, 0.00001);
	QUnit.close(M, 278.99397, 0.00001);
	QUnit.close(C, -1.89732, 0.00001);
	QUnit.close(Lambda, 199.90895, 0.00002);
    QUnit.close(Delta, -7.78507, 0.00001);
    QUnit.close(Alpha, 198.38083, 0.00001);

    // values from Astronomical Algorithms page 88

    jd = adhan.Astronomical.julianDay(1987, 4, 10);
    solar = new adhan.SolarCoordinates(jd);
    T = adhan.Astronomical.julianCentury(jd);

    var Theta0 = adhan.Astronomical.meanSiderealTime(T);
    var Thetaapp = solar.apparentSiderealTime;
    var Omega = adhan.Astronomical.ascendingLunarNodeLongitude(T);
    E0 = adhan.Astronomical.meanObliquityOfTheEcliptic(T);
    L0 = adhan.Astronomical.meanSolarLongitude(T);
    var Lp = adhan.Astronomical.meanLunarLongitude(T);
    var dPsi = adhan.Astronomical.nutationInLongitude(T, L0, Lp, Omega);
    var dE = adhan.Astronomical.nutationInObliquity(T, L0, Lp, Omega);
    var E = E0 + dE;

	QUnit.close(Theta0, 197.693195, 0.000001);
    QUnit.close(Thetaapp, 197.6922295833, 0.0001);

    // values from Astronomical Algorithms page 148

    QUnit.close(Omega, 11.2531, 0.0001);
    QUnit.close(dPsi, -0.0010522,  0.0001);
    QUnit.close(dE, 0.0026230556, 0.00001);
    QUnit.close(E0, 23.4409463889, 0.000001);
    QUnit.close(E, 23.4435694444, 0.00001);
});

QUnit.test("Altitude Of Celestial Body", function(assert) {
	var Phi = 38 + (55 / 60) + (17.0 / 3600);
    var Delta = -6 - (43 / 60) - (11.61 / 3600);
    var H = 64.352133;
    var altitude = adhan.Astronomical.altitudeOfCelestialBody(Phi, Delta, H);
	QUnit.close(altitude, 15.1249, 0.0001);
});

QUnit.test("Transit and Hour Angle", function(assert) {
	// values from Astronomical Algorithms page 103
	var longitude = -71.0833;
	var Theta = 177.74208;
	var Alpha1 = 40.68021;
	var Alpha2 = 41.73129;
	var Alpha3 = 42.78204;
	var m0 = adhan.Astronomical.approximateTransit(longitude, Theta, Alpha2);

	QUnit.close(m0, 0.81965, 0.00001);

    var transit = adhan.Astronomical.correctedTransit(m0, longitude, Theta, Alpha2, Alpha1, Alpha3) / 24;

    QUnit.close(transit, 0.81980, 0.00001);

    var Delta1 = 18.04761;
    var Delta2 = 18.44092;
    var Delta3 = 18.82742;
    var coordinates = new adhan.Coordinates(42.3333, longitude);

    var rise = adhan.Astronomical.correctedHourAngle(m0, -0.5667, coordinates, false, Theta, Alpha2, Alpha1, Alpha3, Delta2, Delta1, Delta3) / 24;
    QUnit.close(rise, 0.51766, 0.00001);

});

QUnit.test("Solar Time", function(assert) {
    /*
    Comparison values generated from http://aa.usno.navy.mil/rstt/onedaytable?form=1&ID=AA&year=2015&month=7&day=12&state=NC&place=raleigh
    */
    var coordinates = new adhan.Coordinates(35 + 47/60, -78 - 39/60);
    var solar = new adhan.SolarTime(new Date(2015, 6, 12), coordinates);

    var transit = solar.transit;
    var sunrise = solar.sunrise;
    var sunset = solar.sunset;
    var twilightStart = solar.hourAngle(-6, false);
    var twilightEnd = solar.hourAngle(-6, true);
    var invalid = solar.hourAngle(-36, true);
    assert.equal(timeString(twilightStart), "9:38");
    assert.equal(timeString(sunrise), "10:08");
    assert.equal(timeString(transit), "17:20");
    assert.equal(timeString(sunset), "24:32");
    assert.equal(timeString(twilightEnd), "25:02");
    assert.equal(timeString(invalid), "");
});

QUnit.test("Right Ascension Edge Case", function(assert) {
    var coordinates = new adhan.Coordinates(35 + 47/60, -78 - 39/60);
    var solar = [];
    for (var i = 0; i <= 365; i++) {
        solar.push(new adhan.SolarTime(new Date(2016, 0, i), coordinates));
    }

    for (var i = 1; i < solar.length; i++) {
        var time = solar[i];
        var previousTime = solar[i-1];
        assert.ok(Math.abs(time.transit - previousTime.transit) < 1/60);
        assert.ok(Math.abs(time.sunrise - previousTime.sunrise) < 2/60);
        assert.ok(Math.abs(time.sunset - previousTime.sunset) < 2/60);
    }
});

QUnit.test("Calendrical Date", function(assert) {
	// generated from http://aa.usno.navy.mil/data/docs/RS_OneYear.php for KUKUIHAELE, HAWAII
    var coordinates = new adhan.Coordinates(20 + 7/60, -155 - 34/60);
    var day1solar = new adhan.SolarTime(new Date(2015, 3, 2), coordinates);
    var day2solar = new adhan.SolarTime(new Date(2015, 3, 3), coordinates);

    var day1 = day1solar.sunrise;
    var day2 = day2solar.sunrise;

    assert.equal(timeString(day1), "16:15")
    assert.equal(timeString(day2), "16:14")
});

QUnit.test("Interpolation", function(assert) {
	var interpolatedValue = adhan.Astronomical.interpolate(0.877366, 0.884226, 0.870531, 4.35/24)
	QUnit.close(interpolatedValue, 0.876125, 0.000001);

    var i1 = adhan.Astronomical.interpolate(1, -1, 3, 0.6);
    QUnit.close(i1, 2.2, 0.000001);
});

QUnit.test("Angle Interpolation", function(assert) {
    var i1 = adhan.Astronomical.interpolateAngles(1, -1, 3, 0.6);
    QUnit.close(i1, 2.2, 0.000001);

    var i2 = adhan.Astronomical.interpolateAngles(1, 359, 3, 0.6);
    QUnit.close(i2, 2.2, 0.000001);
});

QUnit.test("Julian Day", function(assert) {
	/*
	Comparison values generated from http://aa.usno.navy.mil/data/docs/JulianDate.php
	*/
	assert.equal(adhan.Astronomical.julianDay(2010, 1, 2), 2455198.500000);
	assert.equal(adhan.Astronomical.julianDay(2011, 2, 4), 2455596.500000);
	assert.equal(adhan.Astronomical.julianDay(2012, 3, 6), 2455992.500000);
	assert.equal(adhan.Astronomical.julianDay(2013, 4, 8), 2456390.500000);
	assert.equal(adhan.Astronomical.julianDay(2014, 5, 10), 2456787.500000);
	assert.equal(adhan.Astronomical.julianDay(2015, 6, 12), 2457185.500000);
	assert.equal(adhan.Astronomical.julianDay(2016, 7, 14), 2457583.500000);
	assert.equal(adhan.Astronomical.julianDay(2017, 8, 16), 2457981.500000);
	assert.equal(adhan.Astronomical.julianDay(2018, 9, 18), 2458379.500000);
	assert.equal(adhan.Astronomical.julianDay(2019, 10, 20), 2458776.500000);
	assert.equal(adhan.Astronomical.julianDay(2020, 11, 22), 2459175.500000);
	assert.equal(adhan.Astronomical.julianDay(2021, 12, 24), 2459572.500000);

	var jdVal = 2457215.67708333;
	QUnit.close(adhan.Astronomical.julianDay(2015, 7, 12, 4.25), jdVal, 0.000001);

    var date = new Date(2015, 6, 12, 4, 15);
    QUnit.close(julianDate(date), jdVal, 0.000001);

	QUnit.close(adhan.Astronomical.julianDay(2015, 7, 12, 8.0), 2457215.833333, 0.000001);
	QUnit.close(adhan.Astronomical.julianDay(1992, 10, 13, 0.0), 2448908.5, 0.000001);
});

QUnit.test("Julian Hours", function(assert) {
	var j1 = adhan.Astronomical.julianDay(2010, 1, 3);
	var j2 = adhan.Astronomical.julianDay(2010, 1, 1, 48);
	assert.equal(j1, j2);
});

QUnit.test("Leap Year", function(assert) {
	assert.notOk(adhan.Astronomical.isLeapYear(2015));
	assert.ok(adhan.Astronomical.isLeapYear(2016));
	assert.ok(adhan.Astronomical.isLeapYear(1600));
	assert.ok(adhan.Astronomical.isLeapYear(2000));
	assert.ok(adhan.Astronomical.isLeapYear(2400));
	assert.notOk(adhan.Astronomical.isLeapYear(1700));
	assert.notOk(adhan.Astronomical.isLeapYear(1800));
	assert.notOk(adhan.Astronomical.isLeapYear(1900));
	assert.notOk(adhan.Astronomical.isLeapYear(2100));
	assert.notOk(adhan.Astronomical.isLeapYear(2200));
	assert.notOk(adhan.Astronomical.isLeapYear(2300));
	assert.notOk(adhan.Astronomical.isLeapYear(2500));
	assert.notOk(adhan.Astronomical.isLeapYear(2600));
});

QUnit.test("Day of Year", function(assert) {
	assert.equal(1, dayOfYear(new Date(2015, 0, 1)));
	assert.equal(365, dayOfYear(new Date(2015, 11, 31)));
	assert.equal(366, dayOfYear(new Date(2016, 11, 31)));
	assert.equal(32, dayOfYear(new Date(2015, 1, 1)));
});

QUnit.test("Days since solstice", function(assert) {
	assert.equal(11, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 0, 1)), 2016, 1));
	assert.equal(10, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2015, 11, 31)), 2015, 1));
	assert.equal(10, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 11, 31)), 2016, 1));
	assert.equal(0, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 11, 21)), 2016, 1));
	assert.equal(1, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 11, 22)), 2016, 1));
	assert.equal(71, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 2, 1)), 2016, 1));
	assert.equal(70, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2015, 2, 1)), 2015, 1));
	assert.equal(365, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 11, 20)), 2016, 1));
	assert.equal(364, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2015, 11, 20)), 2015, 1));

	assert.equal(0, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2015, 5, 21)), 2015, -1));
	assert.equal(0, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 5, 21)), 2016, -1));
	assert.equal(364, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2015, 5, 20)), 2015, -1));
	assert.equal(365, adhan.Astronomical.daysSinceSolstice(dayOfYear(new Date(2016, 5, 20)), 2016, -1));
});
